Plugin Support, Tracing, and New Test Framework

XerTheSquirrel opened the pull request:
Okay! I will get something important down first that this **breaks backwards compatibility**. Code written using new versions will have to correct their code to handle the changes, but it should be quite simple to migrate existing code no problem.

So this is quite a big one but the important parts of it are:

 * Custom Metrics
 * Performance Entries
 * Plugins
   * Trace Plugin (which is included and enabled by default)
 * Test Framework Rewrite
   * The old testing code is quite ugly compared to this new code.
   * It can test more areas of the code.
   * Adding a new test is as simple as making a new class and adding to the set of tests.

I rewrote the test system because it was quite nasty to be honest, it just felt like adding tests for plugins would end up being a giant mess.

Note that it does not include #17 (the runtime variables) as it has not been pulled in yet, but Git should handle the fast forwarding and such

I would be happy to answer anything about this!

pselle commented:
#17 has been merged, possibly since this was made, and there's a conflict that needs resolving if you can check it out @XerTheSquirrel

pselle commented on PLUGINS.md:
```diff
@@ -0,0 +1,69 @@
+
+All plugins are referred to by their execution state, which is an
+implementation of the class `com.iopipe.plugin.IOpipePluginExecution`.
+To obtain the execution state of a plugin
```
This appears disjointed - is something missing?

XerTheSquirrel replied:
I removed the truncated sentence.

pselle commented on PLUGINS.md:
````diff
@@ -0,0 +1,69 @@
+available plugin):
+
+```
+import com.iopipe.examples.ExampleExecution;
````
I'd like to see these usage instructions more direct in terms of "here is what you need and what you do" so that it's quick for someone using a plugin to understand what's necessary. As is, the lead-in makes this difficult to understand.

XerTheSquirrel replied:
I adjusted the documentation and included a quick usage of plugins.

pselle commented on PLUGINS.md:
```diff
@@ -0,0 +1,69 @@
+
+# Writing Plugins
+
+Plugins utilize the `ServiceLoader` class and therefor means that it is very
```
therefor => therefore

XerTheSquirrel replied:
Fixed this.

pselle commented on README.md:
```diff
@@ -146,3 +176,4 @@ to build.
 * `mvn install`         -- Install the project into your own Maven repository.
 * `mvn site`            -- Generate Maven informational pages.
 * `mvn javadoc:javadoc` -- Generate JavaDoc.
+
```
README is missing info on how to use custom metrics (core service) and would benefit from a link to tracing plugin docs

XerTheSquirrel replied:
I added this documentation.

pselle commented on example/src/main/java/com/iopipe/examples/ExampleExecution.java:
```diff
@@ -0,0 +1,54 @@
+import com.iopipe.plugin.IOpipePluginExecution;
+
+/**
+ * This is an execution for the plugin.
```
Where/why would someone create an execution? (since this is in the examples dir)

XerTheSquirrel replied:
I clarified the use of this class.

pselle commented on example/src/main/java/com/iopipe/examples/Hello.java:
```diff
@@ -17,8 +18,17 @@
+		String __input)
 	{
+		// Send a message to the example plugin
+		__exec.<ExampleExecution>plugin(ExampleExecution.class, (__s) ->
```
Is there an example using custom metrics?

XerTheSquirrel replied:
Yes, there is an example.

pselle commented:
Done with the first read, looking forward to edits.

XerTheSquirrel commented:
Corrected the merge conflict, looking at the other parts.

XerTheSquirrel commented on src/main/java/com/iopipe/CustomMetric.java:
```diff
@@ -0,0 +1,239 @@
+	/**
+	 * Initializes the custom metric with a long value.
+	 *
+	 * @param __name The matric name.
```
Corrected these!

XerTheSquirrel commented on src/main/java/com/iopipe/CustomMetric.java:
```diff
@@ -0,0 +1,239 @@
+	}
+	
+	/**
+	 * Initializes the custom metric with a string and long value.
```
I removed this extra constructor.

XerTheSquirrel commented on src/main/java/com/iopipe/IOpipeMeasurement.java:
```diff
@@ -279,6 +379,87 @@ public RemoteRequest buildRequest()
+					gen.write("name", Objects.toString(p.name(), ""));
+					gen.write("version", Objects.toString(p.version(), ""));
+					gen.write("homepage", Objects.toString(p.homepage(), ""));
+					gen.write("enabled", true);
```
Yes, but I can make it so that disabled plugins can still be tracked.

XerTheSquirrel replied:
My upcoming commit will show plugins which are disabled also now.

XerTheSquirrel commented on src/main/java/com/iopipe/CustomMetric.java:
```diff
@@ -0,0 +1,239 @@
+	}
+	
+	/**
+	 * Does this have a double value?
```
Corrected this and other instances.

XerTheSquirrel commented:
To answer your question "One question: should some of these new classes be organized into folders, to make it easier to navigate the project?".

The plugin classes are within their own package.

The two new interfaces `CustomMetric` and `PerformanceEntry` are in the main package because they are for the most part directly used by `IOpipeMeasurement` and putting them in another package would be a bit confusing.

XerTheSquirrel commented:
Okay so I believe I have handled all of the responses up to this point. In `IOpipeService` I split off the plugin managing code into a class called `__Plugins__` so that should be much easier to read and maintain rather than the code which I had before. I added a utility class for tracing to make it easier to use. If I have missed anything, the commit messages should describe what I have done.

I will await more reviewing!

XerTheSquirrel commented:
As per Corey on Slack, I made corrections to the performance entry `startTime` and `duration` so that the nanosecond precision is not lost.

pselle commented:
Requested @nodebotanist as a reviewer for more docs feedback, can also edit docs further too after this is merged.

pselle commented on example/src/main/java/com/iopipe/examples/Lowercase.java:
```diff
@@ -1,10 +1,11 @@
 package com.iopipe.examples;
 
+import com.amazonaws.services.lambda.runtime.Context;
```
Is this context import used?

pselle commented:
Deleted a remaining comment because it wasn't correct, approving! @nodebotanist would appreciate your thoughts/issues on the merged product, probably easier than navigating the big PR I suspect :D
